src/plugins/discourse/models.js

```diff
@@ -0,0 +1,31 @@
+// @flow
+
+export opaque type ServerUrl: string = string;
+
+/**
+ * Does a strict validation of ServerUrl inputs and returns an opaque type
+ * indicating it matches our assumptions elsewhere.
```

Comment:
These assumptions (which?) need to be documented on the opaque type
itself. Otherwise, the type is not useful. Some questions that I have:

  - The code restricts some portions of the URL (scheme, fragment, etc.)
    but not others (username, password). Is it legal for `ServerUrl`s to
    include these components?
  - If two inputs to `validateServerUrl` return the same output, are the
    inputs to be considered the same server URL? If so, this function
    should be called `parse*`, not `validate*` (as it probably should
    be in any case, just due to its type signature). Specifically:
      - Are `http://example.com` and `http://example.com:80` the same
        server URL? [The `URL` constructor elides canonical ports][cp].
      - Are input URLs permitted to have an optional trailing slash?
      - Are input URLs case-sensitive? Many portions of the URL are
        lowercased by the URL constructor, including protocol and host,
        but username and password are not.

The semantics of this type affect how I’m supposed to read, for
instance, the contract of the `cacheKey` function.

[cp]: https://nodejs.org/api/url.html#url_url_port

Reply:
You're right, so I changed this from a blacklisting to a whitelisting strategy. Sacrificing some error message clarity for easier checking what the assumptions are.

Also updated the naming and comments to use `parse*` and consider the opaque version a normalized URL.